workflows/stairway.py

```diff
@@ -57,43 +57,83 @@ def ts_to_stairway(self, ts_path, num_bootstraps=1, mask_file=None):
+                    for md in m.metadata["mutation_list"]:
+                        mt.append(md["mutation_type"])
+                        if set(mut_types[md["mutation_type"]]) == set("neutral"):
+                            neu_positions.append(m.site)
+                        if set(mut_types[md["mutation_type"]]) == set("non_neutral"):
+                            non_neu_positions.append(m.site)
+                site_class[j] = mut_types[mt[0]] if len(mt) == 1 else "double_hit" 
+            assert sum(site_class == None) == 0
+
+            # All SNP locs
+            snp_locs_all =  [int(x.site.position) for x in ts.variants()]
+            snp_locs = [snp_locs_all[index] for index in neu_positions]
+            snp_locs_non_neutral = [snp_locs_all[index] for index in non_neu_positions]
 
             SFSs = []
-            # Masking
-            # print("This is the number of mutations")
-            # print(ts.get_num_mutations())
-            # print("This is the number of variants")
-            # print(ts.variants())
-            # print("This is the number of sites")
-            # print(ts.num_sites)
-            #retain = np.full(ts.get_num_mutations(), False)
-
-            # This part is not really working (dimension problem)
-            # Masking
-            retain = np.full(ts.num_sites, False)
-            # if mask_file:
-            #     mask_table = pd.read_csv(mask_file, sep="\t", header=None)
-            #     chrom = ts_p.split("/")[-1].split(".")[0]
-            #     sub = mask_table[mask_table[0] == chrom]
-            #     mask_ints = pd.IntervalIndex.from_arrays(sub[1], sub[2])
-            #     snp_locs = [int(x.site.position) for x in ts.variants()]
-            #     tmp_bool = [mask_ints.contains(x) for x in snp_locs]
-            #     retain = np.logical_or(retain, tmp_bool)
-            #     total_length -= np.sum(mask_ints.length)
-
-            #retain = retain.flatten()
-            retain = np.logical_not(retain)
-            # print("This is the shape of retain")
-            # print(retain.shape)
-            # append unmasked SFS
-            SFSs.append(allel.sfs(allel.HaplotypeArray(haps).count_alleles()[:, 1])[1:])
+
+            # Make it work, get ride off neutral positions that overlap the mask region
+            retain = np.full(len(neu_positions), True)
+            retain_non_neu = np.full(len(non_neu_positions), True)
+            if mask_file:
+                mask_table = pd.read_csv(mask_file, sep="\t", header=None)
+                chrom = ts_p.split("/")[-1].split(".")[0]
+                chrom = chrom.split("sim_")[1]
+                sub = mask_table[mask_table[0] == chrom]
+                mask_ints = pd.IntervalIndex.from_arrays(sub[1], sub[2])
+                tmp_bool_neu = [mask_ints.contains(x) for x in snp_locs]
+                tmp_bool_n_neu = [mask_ints.contains(x) for x in snp_locs_non_neutral]
+                for i in range(len(tmp_bool_neu)):
```

Comment:
I think you don't have to mask neutral and non-neutral sites separately. instead use one array to keep track of the masking and another to assign neutral / non-neutral status